### PR TITLE
Jane Street packages v0.12.

### PR DESCRIPTION
**Pull request by xclerc**

### CODE REVIEW

File: packages/config_h/config_h.v0.12.0/opam [hunk 1]
```diff
@@ -0,0 +1,30 @@
+opam-version: "2.0"
```
Comment (avsm) [inline packages/config_h/config_h.v0.12.0/opam]: This package has a fairly generic name -- if it's intended to be more internal, perhaps preface it with `js-config`?

**Comment by avsm** [comment]
Thanks for the contributions! A quick couple of notes on the format of the opam files:

- remove the 0 length `descr` files
- considering using triple quotes around the `description` fields (`"""`) to make them more robust to not needing escaping in the descriptions.

**Comment by xclerc** [comment]
Indeed, we should either get rid of empty descriptions,
or require people to provide one (for some reason, the
former seems a tad more likely).

Triple quoting also sounds like a good idea (even
though all these files are generated, so we can ensure
that strings are properly escaped), but I would have
one question: is there an escape sequence for `"""`?

When should I push such changes? (as in "won't it
restart all checks?")

**Comment by avsm** [comment]
base fails, so push the metadata changes but with a fix for that?

```

#=== ERROR while compiling base.v0.12.0 =======================================#
# context              2.0.3 | linux/x86_64 | ocaml-base-compiler.4.07.1 | git+file:///home/opam/opam-repository
# path                 ~/.opam/4.07/.opam-switch/build/base.v0.12.0
# command              ~/.opam/4.07/bin/dune build -p base -j 71
# exit-code            1
# env-file             ~/.opam/log/base-20-d8fe7b.env
# output-file          ~/.opam/log/base-20-d8fe7b.out
### output ###
# File "shadow-stdlib/src/dune", line 4, characters 0-222:
# 4 | (rule (targets shadow_stdlib.mli)
# 5 |  (deps (:first_dep ../gen/gen.exe)
# 6 |   ../../compiler-stdlib/src/.caml.objs/caml.cmi)
# 7 |  (action
# 8 |   (run %{first_dep} -caml-cmi ../../compiler-stdlib/src/.caml.objs/caml.cmi
# 9 |    -o %{targets})))
# Error: No rule found for compiler-stdlib/src/.caml.objs/caml.cmi
#           sh src/mpopcnt.sexp,src/popcnt_test.c
# /bin/sh: 1: cannot create : Directory nonexistent
```

regarding descriptions, I have a strong preference for at least a one liner -- otherwise we end up with tons of poorly described packages.  Incidentally, you do not need the zero-length `descr` files that are currently part of this PR: the `description:` field is sufficient.  I dont think there is a way to escape `"""`.

**Comment by xclerc** [comment]
It looks like the Base problem is actually a dune problem.
Unless I am mistaken, it builds with dune 1.6.3, but fails
with both 1.7.0 and 1.7.1.

**Comment by xclerc** [comment]
(Please do not merge; I did not rename `config_h`,
but will if everything goes smoothly.)

**Comment by xclerc** [comment]
Am I correct in assuming that the only pending change
is the renaming of the `config_h` package (probably to
`jst-config`)?

**Comment by kit-ty-kate** [comment]
I don't understand the current issue. Why is this making all those packages uninstallable?

**Comment by xclerc** [comment]
I have renamed the `config_h` package to `jst-config`.

@kit-ty-kate Sorry, your message crossed my push.

**Comment by xclerc** [comment]
I am slightly confused; `ci/datakit/1` (at least) seems
to indicate that the package can be built, so why/how do
camelus see them as "broken"?

**Comment by kit-ty-kate** [comment]
I've noticed. I'm currently rebasing your branch on top of master. Bear with me

**Comment by kit-ty-kate** [comment]
There are still a lot of uninstallable packages

**Comment by mseri** [comment]
Can you also move the content of the url files back into the opam files?

**Comment by xclerc** [comment]
I am a bit lost; how can I access the logs of the camelus jobs
to understand the issue(s)?

Note: the Fedora error is an error while retrieving the Docker
image, and the Travis error looks like a time out.

**Comment by xclerc** [comment]
@mseri Will push the requested change alongside a fix for
the "broken packages" issue, once I understand the problem.

**Comment by xclerc** [comment]
OK, I think I understand the problem: we need to relax the
upper bounds of the packages in the dependency cone of
the packages we are trying to push. Commit https://github.com/ocaml/opam-repository/pull/13452/commits/4a8d702312b1ff424b51b57ec58f7e2a3262bf25 is
expected to do that, and commit https://github.com/ocaml/opam-repository/pull/13452/commits/5afef2093dfd0483b9b3937a3134fb4bd4a864d1 addresses the
issue raised by @mseri.

**Comment by kit-ty-kate** [comment]
Something is missing in core:
```
#=== ERROR while compiling core.v0.12.0 =======================================#
# context              2.0.3 | linux/x86_64 | ocaml-base-compiler.4.07.1 | git+file:///home/opam/opam-repository
# path                 ~/.opam/4.07/.opam-switch/build/core.v0.12.0
# command              ~/.opam/4.07/bin/dune build -p core -j 71
# exit-code            1
# env-file             ~/.opam/log/core-20-d3ca31.env
# output-file          ~/.opam/log/core-20-d3ca31.out
### output ###
#       ocamlc src/linux_ext_stubs.o (exit 2)
# (cd _build/default/src && /home/opam/.opam/4.07/bin/ocamlc.opt -g -I /home/opam/.opam/4.07/lib/base -I /home/opam/.opam/4.07/lib/base/caml -I /home/opam/.opam/4.07/lib/base/md5 -I /home/opam/.opam/4.07/lib/base/shadow_stdlib -I /home/opam/.opam/4.07/lib/base_bigstring -I /home/opam/.opam/4.07/lib/base_quickcheck -I /home/opam/.opam/4.07/lib/bin_prot -I /home/opam/.opam/4.07/lib/bin_prot/shape -I /home/opam/.opam/4.07/lib/core_kernel -I /home/opam/.opam/4.07/lib/core_kernel/base_for_tests -I /home/opam/.opam/4.07/lib/fieldslib -I /home/opam/.opam/4.07/lib/jane-street-headers -I /home/opam/.opam/4.07/lib/ocaml/threads -I /home/opam/.opam/4.07/lib/parsexp -I /home/opam/.opam/4.07/lib/ppx_assert/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_bench/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_compare/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_enumerate/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_expect/collector -I /home/opam/.opam/4.07/lib/ppx_expect/common -I /home/opam/.opam/4.07/lib/ppx_expect/config -I /home/opam/.opam/4.07/lib/ppx_hash/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_inline_test/config -I /home/opam/.opam/4.07/lib/ppx_inline_test/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_module_timer/runtime -I /home/opam/.opam/4.07/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.07/lib/sexplib -I /home/opam/.opam/4.07/lib/sexplib/unix -I /home/opam/.opam/4.07/lib/sexplib0 -I /home/opam/.opam/4.07/lib/spawn -I /home/opam/.opam/4.07/lib/splittable_random -I /home/opam/.opam/4.07/lib/stdio -I /home/opam/.opam/4.07/lib/time_now -I /home/opam/.opam/4.07/lib/typerep -I /home/opam/.opam/4.07/lib/variantslib -ccopt -g -ccopt -D_LARGEFILE64_SOURCE -ccopt -I../../core_kernel/src -o linux_ext_stubs.o linux_ext_stubs.c)
# linux_ext_stubs.c:26:24: fatal error: attr/xattr.h: No such file or directory
#  #include <attr/xattr.h>
#                         ^
# compilation terminated.
#     ocamlopt src/.core.objs/native/core__Iobuf.{cmx,o}
# File "src/iobuf.ml", line 1080, characters 4-62:
# Warning 55: Cannot inline: Function information unavailable
```

**Comment by xclerc** [comment]
@kit-ty-kate I'm on it. From what I see so far, the path might
differ according to the OS flavor.

**Comment by xclerc** [comment]
In case there are Linux experts listening, I am contemplating
the following patch:

```
diff --git a/lib/core/src/linux_ext_stubs.c b/lib/core/src/linux_ext_stubs.c
26c26,29
< #include <attr/xattr.h>
---
> #include <sys/xattr.h>
> #ifndef ENOATTR
> # define ENOATTR ENODATA
> #endif
```

**Comment by kevinji** [comment]
Related: janestreet/core#123

**Comment by xclerc** [comment]
@kevinji @edwintorok Indeed; I should have referenced
this PR, the patch above might be marginally better by
using a header from glibc rather than one coming from
another package, but I am no expert.

**Comment by avsm** [comment]
All of this almost certainly breaks on the BSDs, where `attr/xattr.h` doesnt exist. I'll do a manual test of this PR on freebsd shortly...

**Comment by xclerc** [comment]
I had a quick look at the BSDs (well, macOS, to be honest);
they do have a `sys/xattr.h`, **but** the function prototypes
are slightly different in order to support resource forks.
(I thought they were a thing from the past.)

That being said, I have to double-check but the header
inclusion is part of the "linux extensions" of Core, so we
should not even try to compile it when on BSD. (The
paragraph above merely indicates that the support for
extended attributes might be added to BSDs, at the
price of an `#idfef`.)

**Comment by xclerc** [comment]
The `#include` is indeed protected by `#ifdef`s preventing
compilation on non-Linux platform. The problem is likely
to come from Alpine / musl libc, where function prototypes
are also different (*e.g.* no `flags` parameter to "set"
functions).

**Comment by avsm** [comment]
Core fails to compile on FreeBSD:

```
#=== ERROR while compiling core.v0.12.0 =======================================#                                                                                                       
# context     2.0.1 | freebsd/x86_64 | ocaml-base-compiler.4.07.1 | git+file:///root/scratch/opam-repository                                                                           
# path        ~/.opam/4.07.1/.opam-switch/build/core.v0.12.0                                                                                                                           
# command     ~/.opam/4.07.1/bin/dune build -p core -j 3                                                                                                                               
# exit-code   1                                                                                                                                                                        
# env-file    ~/.opam/log/core-7120-71abc2.env                                                                                                                                         
# output-file ~/.opam/log/core-7120-71abc2.out                                                                                                                                         
### output ###                                                                                                                                                                         
# Error: The implementation src/linux_ext.pp.ml                                                                                                                                        
# [...]                                                                                                                                                                                
#        The module `Extended_file_attributes' is required but not provided                                                                                                            
#        File "src/linux_ext_intf.ml", line 499, characters 2-2635:                                                                                                                    
#          Expected declaration                                                                                                                                                        
#     ocamlopt src/.core.objs/native/core__Linux_ext.{cmx,o} (exit 2)                                                                                                                  
# (cd _build/default && /root/.opam/4.07.1/bin/ocamlopt.opt -w -40 -g -I src/.core.objs/byte -I src/.core.objs/native -I /root/.opam/4.07.1/lib/base -I /root/.opam/4.07.1/lib/base/cam
l -I /root/.opam/4.07.1/lib/base/md5 -I /root/.opam/4.07.1/lib/base/shadow_stdlib -I /root/.opam/4.07.1/lib/base_bigstring -I /root/.opam/4.07.1/lib/base_quickcheck -I /root/.opam/4.0
7.1/lib/bin_prot -I /root/.opam[...]                                                                                                                                                   
# File "src/linux_ext.ml", line 1:                                                                                                                                                     
# Error: The implementation src/linux_ext.pp.ml                                                                                                                                        
#        does not match the interface src/.core.objs/byte/core__Linux_ext.cmi:                                                                                                         
#        The module `Extended_file_attributes' is required but not provided                                                                                                            
#        File "src/linux_ext_intf.ml", line 499, characters 2-2635:                                                                                                                    
#          Expected declaration                                                                                                                                                        
                                                                                                                                                                                       ```

**Comment by avsm** [comment]
I wont have time to debug the freebsd issue today unfortunately, but ping me if you need an account on a freebsd machine to debug.

**Comment by xclerc** [comment]
@asvm Thanks; I have been able to reproduce the problem
on macOS, and will work under the assumption that FreeBSD
and macOS are equivalent (that is, for the problem at hand).

**Comment by avsm** [comment]
sgtm, I can re-test pretty quickly on freebsd if you ping me when ready

**Comment by xclerc** [comment]
Well, for core, the patch below seems to be enough.
However, do not waste your time: I still have to check the other packages;
will ping you once everything is fixed on my box.

```
--- a/src/linux_ext.ml
+++ b/src/linux_ext.ml
@@ -1084,5 +1084,4 @@ end
 [%%else]
 include Null_toplevel
 module Eventfd = Null.Eventfd
+module Extended_file_attributes = Null.Extended_file_attributes
 [%%endif]
```

**Comment by avsm** [comment]
Btw you might want to crib from avsm/ocaml-yaml on github that has a Azure Pipelines CI script that'll test a dune package on Windows, macOS and Linux simultaneously. Just built it a few days ago and it's convenient to have all the tests in one provider.

**Comment by xclerc** [comment]
The only package having issues on macOS is `pam` (I have
a patch that fixes the build, but it needs further investigation);
`netsnmp` is not built, but it is expected.

Anyway, it looks like `pam` needs a `conf-pam` companion
package, an we can disable macOS support there.

**Comment by kit-ty-kate** [comment]
core seems to compile so far

**Comment by xclerc** [comment]
`pam` is failing, which is more or less expected, and the error with
base should be fixed by https://github.com/ocaml/opam-repository/pull/13452/commits/7ccd27037ca7141eb93e99c420870b9be2774864.

**Comment by kit-ty-kate** [comment]
Could you create a new `conf-libpam` package which would pull at least `libpam0g-dev` on debian? Fix `pam` depends on that package, that should fix the issue I suppose

**Comment by xclerc** [comment]
I concur; however, a quick experiment shows that
the package is not registered with `pkg-config`.
Is that a problem?

**Comment by kit-ty-kate** [comment]
that's not problem. You can instead test whether a C compiler can link with the library.
For example see: https://github.com/ocaml/opam-repository/tree/master/packages/conf-rocksdb/conf-rocksdb.1

**Comment by xclerc** [comment]
@kit-ty-kate Thanks for the pointer!
I have pushed a similar configuration package for pam;
hopefully, everything will be green.

**Comment by xclerc** [comment]
Please do not merge even if everything is green;
I will push a fix for https://github.com/janestreet/base/issues/67.

**Comment by xclerc** [comment]
https://github.com/ocaml/opam-repository/pull/13452/commits/b5783e257361794adef49a4a1b0a7953a9461299 is expected to fix https://github.com/janestreet/base/issues/67.

(During the previous run, the CI had a number of errors because
I deleted and then re-generated the archive for `base` and `jst-config`.)

**Comment by avsm** [comment]
This looks good on freebsd, except this query in conf-pam:

```
+ /usr/bin/cc "-lpam" "main.c" (CWD=/root/.opam/4.07.1/.opam-switch/build/conf-pam.1)
- main.c:3:29: error: use of undeclared identifier 'PAM_XDISPLAY'
- static int missing_on_bsd = PAM_XDISPLAY;
-                             ^
```

if we have deliberate the `missing_on_bsd` check for the symbol, then setting an `os = linux` seems wise in this package to avoid it getting selected on freebsd. However, there is a `libpam` installed in freebsd base, so this seems a bit harsh...

**Comment by avsm** [comment]
Almost there. The last failure seems to be in musl-libc as found in Alpine Linux, which would be good to fix:

```
#=== ERROR while compiling core.v0.12.0 =======================================#
# context              2.0.3 | linux/x86_64 | ocaml-base-compiler.4.07.1 | git+file:///home/opam/opam-repository
# path                 ~/.opam/4.07/.opam-switch/build/core.v0.12.0
# command              ~/.opam/4.07/bin/dune build -p core -j 71
# exit-code            1
# env-file             ~/.opam/log/core-18-d3ca31.env
# output-file          ~/.opam/log/core-18-d3ca31.out
### output ###
#       ocamlc src/linux_ext_stubs.o (exit 2)
# (cd _build/default/src && /home/opam/.opam/4.07/bin/ocamlc.opt -g -I /home/opam/.opam/4.07/lib/base -I /home/opam/.opam/4.07/lib/base/caml -I /home/opam/.opam/4.07/lib/base/md5 -I /home/opam/.opam/4.07/lib/base/shadow_stdlib -I /home/opam/.opam/4.07/lib/base_bigstring -I /home/opam/.opam/4.07/lib/base_quickcheck -I /home/opam/.opam/4.07/lib/bin_prot -I /home/opam/.opam/4.07/lib/bin_prot/shape -I /home/opam/.opam/4.07/lib/core_kernel -I /home/opam/.opam/4.07/lib/core_kernel/base_for_tests -I /home/opam/.opam/4.07/lib/fieldslib -I /home/opam/.opam/4.07/lib/jane-street-headers -I /home/opam/.opam/4.07/lib/ocaml/threads -I /home/opam/.opam/4.07/lib/parsexp -I /home/opam/.opam/4.07/lib/ppx_assert/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_bench/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_compare/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_enumerate/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_expect/collector -I /home/opam/.opam/4.07/lib/ppx_expect/common -I /home/opam/.opam/4.07/lib/ppx_expect/config -I /home/opam/.opam/4.07/lib/ppx_hash/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_inline_test/config -I /home/opam/.opam/4.07/lib/ppx_inline_test/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_module_timer/runtime -I /home/opam/.opam/4.07/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.07/lib/sexplib -I /home/opam/.opam/4.07/lib/sexplib/unix -I /home/opam/.opam/4.07/lib/sexplib0 -I /home/opam/.opam/4.07/lib/spawn -I /home/opam/.opam/4.07/lib/splittable_random -I /home/opam/.opam/4.07/lib/stdio -I /home/opam/.opam/4.07/lib/time_now -I /home/opam/.opam/4.07/lib/typerep -I /home/opam/.opam/4.07/lib/variantslib -ccopt -g -ccopt -D_LARGEFILE64_SOURCE -ccopt -I../../core_kernel/src -o linux_ext_stubs.o linux_ext_stubs.c)
# linux_ext_stubs.c: In function 'linux_sendmsg_nonblocking_no_sigpipe_stub':
# linux_ext_stubs.c:154:46: warning: initialization of 'int' from 'void *' makes integer from pointer without a cast [-Wint-conversion]
#    struct msghdr msghdr = { NULL, 0, NULL, 0, NULL, 0, 0 };
#                                               ^~~~
# linux_ext_stubs.c:154:46: note: (near initialization for 'msghdr.__pad1')
# linux_ext_stubs.c: In function 'linux_getxattr':
# linux_ext_stubs.c:648:12: error: 'XATTR_SIZE_MAX' undeclared (first use in this function); did you mean 'SSIZE_MAX'?
#    char buf[XATTR_SIZE_MAX + 1];
#             ^~~~~~~~~~~~~~
#             SSIZE_MAX
# linux_ext_stubs.c:648:12: note: each undeclared identifier is reported only once for each function it appears in
#       ocamlc src/bigstring_stubs.o
# bigstring_stubs.c:39: warning: "__BYTE_ORDER" redefined
#  #define __BYTE_ORDER    _BYTE_ORDER
#  
# In file included from /usr/include/sys/types.h:70,
#                  from bigstring_stubs.c:33:
# /usr/include/endian.h:11: note: this is the location of the previous definition
#  #define __BYTE_ORDER __BYTE_ORDER__
#  
# bigstring_stubs.c:40: warning: "__LITTLE_ENDIAN" redefined
#  #define __LITTLE_ENDIAN _LITTLE_ENDIAN
#  
# In file included from /usr/include/sys/types.h:70,
#                  from bigstring_stubs.c:33:
# /usr/include/endian.h:6: note: this is the location of the previous definition
#  #define __LITTLE_ENDIAN 1234
#  
# bigstring_stubs.c:41: warning: "__BIG_ENDIAN" redefined
#  #define __BIG_ENDIAN    _BIG_ENDIAN
#  
# In file included from /usr/include/sys/types.h:70,
#                  from bigstring_stubs.c:33:
# /usr/include/endian.h:7: note: this is the location of the previous definition
#  #define __BIG_ENDIAN 4321
#  
# bigstring_stubs.c: In function 'bigstring_sendmsg_nonblocking_no_sigpipe_stub':
# bigstring_stubs.c:713:46: warning: initialization of 'int' from 'void *' makes integer from pointer without a cast [-Wint-conversion]
#    struct msghdr msghdr = { NULL, 0, NULL, 0, NULL, 0, 0 };
#                                               ^~~~
# bigstring_stubs.c:713:46: note: (near initialization for 'msghdr.__pad1')
#       ocamlc src/unix_stubs.o
# unix_stubs.c:350:2: warning: #warning "_POSIX_SYNCHRONIZED_IO undefined or <= 0; aliasing unix_fdatasync to unix_fsync" [-Wcpp]
#  #warning "_POSIX_SYNCHRONIZED_IO undefined or <= 0; aliasing unix_fdatasync to unix_fsync"
#   ^~~~~~~
# unix_stubs.c:1367:2: warning: #warning "_POSIX_PRIORITY_SCHEDULING not present; sched_setscheduler undefined" [-Wcpp]
#  #warning "_POSIX_PRIORITY_SCHEDULING not present; sched_setscheduler undefined"
#   ^~~~~~~
#     ocamlopt src/.core.objs/native/core__Iobuf.{cmx,o}
# File "src/iobuf.ml", line 1080, characters 4-62:
# Warning 55: Cannot inline: Function information unavailable
```

Also Warning 55 there looks worrying -- is that a separate problem?

**Comment by xclerc** [comment]
@avsm Thanks! I will have a look at both issues.
Two libraries have also been unfortunately dropped,
and will add them back.

**Comment by avsm** [comment]
if you depend on limits.h in musl, you now need to add the alpine `linux-headers` depext to core v0.12.0.  note there are still other compilation bugs with alpine in the new push:

```
#=== ERROR while compiling core.v0.12.0 =======================================#
# context              2.0.3 | linux/x86_64 | ocaml-base-compiler.4.07.1 | git+file:///home/opam/opam-repository
# path                 ~/.opam/4.07/.opam-switch/build/core.v0.12.0
# command              ~/.opam/4.07/bin/dune build -p core -j 71
# exit-code            1
# env-file             ~/.opam/log/core-19-9d0fb6.env
# output-file          ~/.opam/log/core-19-9d0fb6.out
### output ###
#       ocamlc src/linux_ext_stubs.o (exit 2)
# (cd _build/default/src && /home/opam/.opam/4.07/bin/ocamlc.opt -g -I /home/opam/.opam/4.07/lib/base -I /home/opam/.opam/4.07/lib/base/caml -I /home/opam/.opam/4.07/lib/base/md5 -I /home/opam/.opam/4.07/lib/base/shadow_stdlib -I /home/opam/.opam/4.07/lib/base_bigstring -I /home/opam/.opam/4.07/lib/base_quickcheck -I /home/opam/.opam/4.07/lib/bin_prot -I /home/opam/.opam/4.07/lib/bin_prot/shape -I /home/opam/.opam/4.07/lib/core_kernel -I /home/opam/.opam/4.07/lib/core_kernel/base_for_tests -I /home/opam/.opam/4.07/lib/fieldslib -I /home/opam/.opam/4.07/lib/jane-street-headers -I /home/opam/.opam/4.07/lib/ocaml/threads -I /home/opam/.opam/4.07/lib/parsexp -I /home/opam/.opam/4.07/lib/ppx_assert/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_bench/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_compare/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_enumerate/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_expect/collector -I /home/opam/.opam/4.07/lib/ppx_expect/common -I /home/opam/.opam/4.07/lib/ppx_expect/config -I /home/opam/.opam/4.07/lib/ppx_hash/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_inline_test/config -I /home/opam/.opam/4.07/lib/ppx_inline_test/runtime-lib -I /home/opam/.opam/4.07/lib/ppx_module_timer/runtime -I /home/opam/.opam/4.07/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.07/lib/sexplib -I /home/opam/.opam/4.07/lib/sexplib/unix -I /home/opam/.opam/4.07/lib/sexplib0 -I /home/opam/.opam/4.07/lib/spawn -I /home/opam/.opam/4.07/lib/splittable_random -I /home/opam/.opam/4.07/lib/stdio -I /home/opam/.opam/4.07/lib/time_now -I /home/opam/.opam/4.07/lib/typerep -I /home/opam/.opam/4.07/lib/variantslib -ccopt -g -ccopt -D_LARGEFILE64_SOURCE -ccopt -I../../core_kernel/src -o linux_ext_stubs.o linux_ext_stubs.c)
# linux_ext_stubs.c:33:10: fatal error: linux/limits.h: No such file or directory
#  #include <linux/limits.h> /* needed to build with musl */
#           ^~~~~~~~~~~~~~~~
# compilation terminated.
#       ocamlc src/bigstring_stubs.o
# bigstring_stubs.c:39: warning: "__BYTE_ORDER" redefined
#  #define __BYTE_ORDER    _BYTE_ORDER
#  
# In file included from /usr/include/sys/types.h:70,
#                  from bigstring_stubs.c:33:
# /usr/include/endian.h:11: note: this is the location of the previous definition
#  #define __BYTE_ORDER __BYTE_ORDER__
#  
# bigstring_stubs.c:40: warning: "__LITTLE_ENDIAN" redefined
#  #define __LITTLE_ENDIAN _LITTLE_ENDIAN
#  
# In file included from /usr/include/sys/types.h:70,
#                  from bigstring_stubs.c:33:
# /usr/include/endian.h:6: note: this is the location of the previous definition
#  #define __LITTLE_ENDIAN 1234
#  
# bigstring_stubs.c:41: warning: "__BIG_ENDIAN" redefined
#  #define __BIG_ENDIAN    _BIG_ENDIAN
#  
# In file included from /usr/include/sys/types.h:70,
#                  from bigstring_stubs.c:33:
# /usr/include/endian.h:7: note: this is the location of the previous definition
#  #define __BIG_ENDIAN 4321
#  
# bigstring_stubs.c: In function 'bigstring_sendmsg_nonblocking_no_sigpipe_stub':
# bigstring_stubs.c:713:46: warning: initialization of 'int' from 'void *' makes integer from pointer without a cast [-Wint-conversion]
#    struct msghdr msghdr = { NULL, 0, NULL, 0, NULL, 0, 0 };
#                                               ^~~~
# bigstring_stubs.c:713:46: note: (near initialization for 'msghdr.__pad1')
#       ocamlc src/unix_stubs.o
# unix_stubs.c:350:2: warning: #warning "_POSIX_SYNCHRONIZED_IO undefined or <= 0; aliasing unix_fdatasync to unix_fsync" [-Wcpp]
#  #warning "_POSIX_SYNCHRONIZED_IO undefined or <= 0; aliasing unix_fdatasync to unix_fsync"
#   ^~~~~~~
# unix_stubs.c:1367:2: warning: #warning "_POSIX_PRIORITY_SCHEDULING not present; sched_setscheduler undefined" [-Wcpp]
#  #warning "_POSIX_PRIORITY_SCHEDULING not present; sched_setscheduler undefined"
#   ^~~~~~~
#     ocamlopt src/.core.objs/native/core__Iobuf.{cmx,o}
# File "src/iobuf.ml", line 1080, characters 4-62:
# Warning 55: Cannot inline: Function information unavailable
```

one full log: https://ci.ocamllabs.io/log/saved/docker-build-74df1a1434d85f221ac8332f97266c09/83219b8beda5acb73df704e8d30463c95c92c6bc

**Comment by avsm** [comment]
I think this is mergeable now; marking it as so for when the CI finishes with Fedora.  Just a minor problem with alpine/conf-pam to fix in a subsequent PR:

```
<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
Processing  1/2: [conf-pam: cc main.c]
+ /usr/bin/cc "-lpam" "main.c" (CWD=/home/opam/.opam/4.07/.opam-switch/build/conf-pam.1)
- main.c:1:10: fatal error: security/pam_appl.h: No such file or directory
-  #include <security/pam_appl.h>
-           ^~~~~~~~~~~~~~~~~~~~~
- compilation terminated.
```

**Comment by xclerc** [comment]
Indeed, there is an embarrassing mistake in `conf-pam`:

```
  ["linux-pam-dev"] {os-distribution = "linux-pam-dev"}
```

should obviously be:

```
  ["linux-pam-dev"] {os-distribution = "alpine"}
```

**Comment by xclerc** [comment]
Also, it might be good to re-run the script adding upper bounds
to reverse dependencies. Just in case PRs merged since this one
was opened have added such dependencies.

**Comment by avsm** [comment]
That sounds good; if you want to re-run that upper bound script and submit it as a separate PR, I can do a gang-merge of that, this PR and the minor fix to conf-pam at the same time.

**Comment by xclerc** [comment]
The PR adding the upper bounds is https://github.com/ocaml/opam-repository/pull/13550, and
the one fixing the `conf-pam` issue is https://github.com/ocaml/opam-repository/pull/13551.

So, I guess the merge order is https://github.com/ocaml/opam-repository/pull/13550, https://github.com/ocaml/opam-repository/pull/13452,
and finally https://github.com/ocaml/opam-repository/pull/13551.

**Comment by mseri** [comment]
Should we merge now?

**Comment by xclerc** [comment]
I would say "yes". My previous comment about the merge
order is outdated since https://github.com/ocaml/opam-repository/pull/13452/commits/4ba143010eca8de69aa42b2f66f7729293317f55.

**Comment by kit-ty-kate** [comment]
Thanks!

**Comment by xclerc** [comment]
@avsm @kit-ty-kate 
Many thanks for your help! (and patience...)
I really appreciate it.

**Comment by avsm** [comment]
Thank you for all your hard work @xclerc!  The third part is equally important now though -- we need to do the upper bound relaxation pieces.  Is this something you want to experiment with via opam-check-all @kit-ty-kate?  I posted an update on https://discuss.ocaml.org/t/experimenting-with-a-new-opam-repository-release-strategy-for-large-libraries/2918/21?u=avsm

**Comment by xclerc** [comment]
Sure, please let me know if I can be of any help.

**Comment by kit-ty-kate** [comment]
@xclerc mmh, to test relaxed upper-bound I'd just need a branch of opam-repository without any of the upper-bound constraints

**Comment by xclerc** [comment]
Sorry, I am not sure I understand: you mean a branch with
no upper bounds on Jane Street packages whatsoever?
I thought you would gradually remove such bounds.

**Comment by kit-ty-kate** [comment]
> Sorry, I am not sure I understand: you mean a branch with
> no upper bounds on Jane Street packages whatsoever?

Yes

> I thought you would gradually remove such bounds.

mmh, well I can launch a general check to speed up the process.

**Comment by avsm** [comment]
I think it might be useful to have a general way to do this, so that you can progressively apply it on the JS packages starting from root ones (like ppx_type_conv) and then slowly working up to core and async.

I think that `opam-ed` might be able to do this, or by modifying `opam admin add-constraints` to have a remove-constraints option. pinging @AltGr at retreat might be best :-)

**Pull request merged**